File: backend/app/schemas/progress.py
from datetime import date, datetime
from pydantic import BaseModel

class ProgressSummary(BaseModel):
    workouts_completed: int
    total_volume_kg: float
    volume_change_percent: float | None
    training_time_seconds: int
    

class RecentWorkout(BaseModel):
    session_id: str
    name: str
    completed_at: datetime
    duration_seconds: int
    set_count: int
    volume_kg: float
 
 
class WeeklyVolumePoint(BaseModel):
    week_start: date
    volume_kg: float   
    
    
class ExerciseProgressHistoryPoint(BaseModel):
    date: datetime
    value: float


class ExerciseProgressItem(BaseModel):
    exercise_id: str
    exercise_name: str
    tracking_type: str

    start_value: float
    current_value: float

    change: float
    change_percent: float | None

    unit: str

    history: list[ExerciseProgressHistoryPoint]

    
class ProgressOverviewResponse(BaseModel):
    range: str
    summary: ProgressSummary
    weekly_volume: list[WeeklyVolumePoint]
    exercise_progress: list[ExerciseProgressItem]
    recent_workouts: list[RecentWorkout]




    

File: frontend/types/progress.ts
export type ProgressRange = "7d" | "30d" | "3m" | "6m" | "1y" | "all";

export interface ProgressOverviewResponse {
    range: ProgressRange;

    summary: {
        workouts_completed: number;
        total_volume_kg: number;
        volume_change_percent: number | null;
        training_time_seconds: number;
    };

    weekly_volume: {
        week_start: string;
        volume_kg: number;
    }[];

    exercise_progress: {
        exercise_id: string;
        exercise_name: string;
        tracking_type: string;
        start_value: number;
        current_value: number;
        change: number;
        change_percent: number | null;
        unit: string;
        history: {
            date: string;
            value: number;
        }[];
    }[];

    recent_workouts: {
        session_id: string;
        name: string;
        completed_at: string;
        duration_seconds: number;
        set_count: number;
        volume_kg: number;
    }[];
}

export interface ProgressOverviewRequest {
    accessToken: string;
    range: ProgressRange;
}